pkg/cloudprovider/rhsm/subscription_manager.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright 2019 The Machine Controller Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rhsm
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"k8s.io/klog"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// RedHatSubscriptionManager is responsible for removing redhat subscriptions.
+type RedHatSubscriptionManager interface {
+	UnregisterInstance(machineName string) error
+}
+
+type pagination struct {
+	Offset int `json:"offset"`
+	Limit  int `json:"limit"`
+	Count  int `json:"count"`
+}
+
+type body struct {
+	Name string `json:"name"`
+	UUID string `json:"uuid"`
+}
+
+type systemsResponse struct {
+	Pagination pagination `json:"pagination"`
+	Body       []*body    `json:"body"`
+}
+
+type credentials struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int    `json:"expires_in"`
+}
+
+type defaultRedHatSubscriptionManager struct {
+	offlineToken string
+	authURL      string
+	apiURL       string
+	client       *http.Client
+	credentials  *credentials
+}
+
+var errUnauthenticatedRequest = errors.New("unauthenticated")
+
+func NewRedHatSubscriptionManager(offlineToken string) (RedHatSubscriptionManager, error) {
+	if offlineToken == "" {
+		return nil, errors.New("offline token, authURL, or apiPath cannot be empty")
+	}
+
+	return &defaultRedHatSubscriptionManager{
+		client:       &http.Client{},
+		authURL:      "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token",
+		apiURL:       "https://api.access.redhat.com/management/v1/systems",
+		offlineToken: offlineToken,
+	}, nil
+}
+
+func (d *defaultRedHatSubscriptionManager) UnregisterInstance(machineName string) error {
+	if d.credentials == nil {
+		klog.Info("access token has been expired, refreshing token")
+		if err := d.refreshToken(); err != nil {
+			return fmt.Errorf("failed to refresh offline token")
+		}
+	}
+
+	var (
+		retries    = 0
+		maxRetries = 15
+	)
+
+	for retries < maxRetries {
```

Comment:
this loop needs backoff logic, and probably backoff should be reported to calling functions somehow (I mean the API surface of this package).

Reply:
I am not quite sure that back off mechanism is really necessary here, the reason for that is we try to execute this loop 15 times every two seconds, not 100 times each 500 milliseconds, and if this fails forever( > 15) the reconcile loop will trigger this again on the next reconcile which has a back-off mechanism( which has a back-off mechanism )